jupyterhub/app.py

```diff
@@ -2358,6 +2359,195 @@ async def init_api_tokens(self):
         )
         pc.start()
 
+    def service_from_orm(
+        self,
+        orm_service: orm.Service,
+        domain: str,
+        host: str,
+    ) -> Service:
+        """Create the service instance and related objects from
+        ORM data.
+
+        Args:
+            orm_service (orm.Service): The `orm.Service` object
+            domain (str): The domain of Hub session
+            host (str):  The host of Hub session
+
+        Returns:
+            Service: the created service
+        """
+
+        name = orm_service.name
+        service = Service(
+            parent=self,
+            app=self,
+            base_url=self.base_url,
+            db=self.db,
+            orm=orm_service,
+            roles=orm_service.roles,
+            domain=domain,
+            host=host,
+            hub=self.hub,
+        )
+        traits = service.traits(input=True)
+        for key in traits:
+            value = orm_service.get_column(key)
+            if value is not None:
+                setattr(service, key, value)
+
+        if orm_service.oauth_client is not None:
+            service.oauth_redirect_uri = orm_service.oauth_client.redirect_uri
+            service.oauth_client_id = orm_service.oauth_client.identifier
+            service.oauth_redirect_uri = orm_service.oauth_client.redirect_uri
+
+        self._service_map[name] = service
+
+        return service
+
+    def service_from_spec(
+        self,
+        spec: Dict,
+        from_config=True,
+        domain: Optional[str] = None,
+        host: Optional[str] = None,
+    ) -> Optional[Service]:
+        """Create the service instance and related objects from
+        config data.
+
+        Args:
+            spec (Dict): The spec of service, defined in the config file.
+            from_config (bool, optional): `True` if the service will be created
+            from the config file, `False` if it is created from REST API.
+            Defaults to `True`.
+            domain (Optional[str]): The domain of Hub session. Defaults to None.
+            host (Optional[str]): The host of Hub session. Defaults to None.
+
+        Returns:
+            Optional[Service]: The created service
+        """
+
+        if domain is None:
+            if self.domain:
```

Comment:
This is always False

Reply:
Sorry, I do not catch that. Do you mean the check `if self.domain` is always `False`? I thought that `self.domain` is configurable.

Reply:
Sorry for not clarifying `domain` is defined as domain of the hub, while `self.domain` is also the domain of the hub, so these two variables should have the same value. I don't think there's a reason to have these args, and instead always take the `if domain is None:` path for domain here and host below.